From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ashish Mhetre <amhetre@nvidia.com>
Date: Tue, 12 Aug 2025 06:32:27 +0000
Subject: [PATCH 200/243] NVIDIA: SAUCE: iommu/arm-smmu-v3: Implement WAR for
 Tegra264 Errata T264-SMMU-3

If there is active traffic on a page while it is being invalidated, it
is possible that the intended invalidated entry is not properly evicted
from the SMMU TLB caches. Subsequent requests to the same virtual
address could get the older mapping and access the wrong location, which
can be a functional and a security problem.
Repeat the entire invalidation->sync sequence once for every
invalidation queued. The second set of invalidation + sync commands is
guaranteed to evict the entry out of all TLBs.

Bug 5448714
Bug 5150304

Change-Id: I40e891a1459f4c65e4e2e7f0224867d6c1be277a
(cherry-picked from commit 27c7fb19c438457aabe7d7b4812bf88778750017)
Signed-off-by: Ashish Mhetre <amhetre@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-noble/+/3496080
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-noble/+/3524355
GVS: buildbot_gerritrpt <buildbot_gerritrpt@nvidia.com>
Reviewed-by: Pritesh Raithatha <praithatha@nvidia.com>
(cherry picked from commit 82f92000e9f39443713157dd1f34935e93fff932)

diff --git a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3-sva.c b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3-sva.c
index f1f8e01a7e91..e852dd5e8be3 100644
--- a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3-sva.c
+++ b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3-sva.c
@@ -139,6 +139,9 @@ static void arm_smmu_mm_arch_invalidate_secondary_tlbs(struct mmu_notifier *mn,
 	size = end - start;
 
 	arm_smmu_domain_inv_range(smmu_domain, start, size, PAGE_SIZE, false);
+	if (smmu_domain->smmu->options & ARM_SMMU_OPT_TLBI_TWICE &&
+	    !(smmu_domain->smmu->features & ARM_SMMU_FEAT_BTM))
+		arm_smmu_domain_inv_range(smmu_domain, start, size, PAGE_SIZE, false);
 }
 
 static void arm_smmu_mm_release(struct mmu_notifier *mn, struct mm_struct *mm)
@@ -170,6 +173,8 @@ static void arm_smmu_mm_release(struct mmu_notifier *mn, struct mm_struct *mm)
 	spin_unlock_irqrestore(&smmu_domain->devices_lock, flags);
 
 	arm_smmu_domain_inv(smmu_domain);
+	if (smmu_domain->smmu->options & ARM_SMMU_OPT_TLBI_TWICE)
+		arm_smmu_domain_inv(smmu_domain);
 }
 
 static void arm_smmu_mmu_notifier_free(struct mmu_notifier *mn)
diff --git a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
index 93ee43b12ba8..c53792ec331a 100644
--- a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
+++ b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
@@ -91,6 +91,7 @@ DEFINE_MUTEX(arm_smmu_asid_lock);
 static struct arm_smmu_option_prop arm_smmu_options[] = {
 	{ ARM_SMMU_OPT_SKIP_PREFETCH, "hisilicon,broken-prefetch-cmd" },
 	{ ARM_SMMU_OPT_PAGE0_REGS_ONLY, "cavium,cn9900-broken-page1-regspace"},
+	{ ARM_SMMU_OPT_TLBI_TWICE, "nvidia,tegra264-tlbi-race-fix"},
 	{ 0, NULL},
 };
 
@@ -1507,6 +1508,8 @@ static void arm_smmu_sync_cd(struct arm_smmu_master *master,
 						  leaf));
 
 	arm_smmu_cmdq_batch_submit(smmu, &cmds);
+	if (smmu->options & ARM_SMMU_OPT_TLBI_TWICE)
+		arm_smmu_cmdq_batch_submit(smmu, &cmds);
 }
 
 static void arm_smmu_write_cd_l1_desc(struct arm_smmu_cdtab_l1 *dst,
@@ -1802,6 +1805,10 @@ static void arm_smmu_ste_writer_sync_entry(struct arm_smmu_entry_writer *writer)
 	arm_smmu_cmdq_issue_cmd_with_sync(
 		writer->master->smmu,
 		arm_smmu_make_cmd_cfgi_ste(ste_writer->sid, true));
+	if (writer->master->smmu->options & ARM_SMMU_OPT_TLBI_TWICE)
+		arm_smmu_cmdq_issue_cmd_with_sync(
+			writer->master->smmu,
+			arm_smmu_make_cmd_cfgi_ste(ste_writer->sid, true));
 }
 
 static const struct arm_smmu_entry_writer_ops arm_smmu_ste_writer_ops = {
@@ -2439,6 +2446,8 @@ static void arm_smmu_tlb_inv_context(void *cookie)
 	 * writes from another CPU before it executes the IOTLB invalidation.
 	 */
 	arm_smmu_domain_inv(smmu_domain);
+	if (smmu_domain->smmu->options & ARM_SMMU_OPT_TLBI_TWICE)
+		arm_smmu_domain_inv(smmu_domain);
 }
 
 static void arm_smmu_cmdq_batch_add_range(struct arm_smmu_device *smmu,
@@ -4797,15 +4806,23 @@ static int arm_smmu_device_reset(struct arm_smmu_device *smmu)
 
 	/* Invalidate any cached configuration */
 	arm_smmu_cmdq_issue_cmd_with_sync(smmu, arm_smmu_make_cmd_cfgi_all());
+	if (smmu->options & ARM_SMMU_OPT_TLBI_TWICE)
+		arm_smmu_cmdq_issue_cmd_with_sync(smmu, arm_smmu_make_cmd_cfgi_all());
 
 	/* Invalidate any stale TLB entries */
 	if (smmu->features & ARM_SMMU_FEAT_HYP) {
 		arm_smmu_cmdq_issue_cmd_with_sync(
 			smmu, arm_smmu_make_cmd_op(CMDQ_OP_TLBI_EL2_ALL));
+		if (smmu->options & ARM_SMMU_OPT_TLBI_TWICE)
+			arm_smmu_cmdq_issue_cmd_with_sync(
+				smmu, arm_smmu_make_cmd_op(CMDQ_OP_TLBI_EL2_ALL));
 	}
 
 	arm_smmu_cmdq_issue_cmd_with_sync(
 		smmu, arm_smmu_make_cmd_op(CMDQ_OP_TLBI_NSNH_ALL));
+	if (smmu->options & ARM_SMMU_OPT_TLBI_TWICE)
+		arm_smmu_cmdq_issue_cmd_with_sync(
+			smmu, arm_smmu_make_cmd_op(CMDQ_OP_TLBI_NSNH_ALL));
 
 	/* Event queue */
 	writeq_relaxed(smmu->evtq.q.q_base, smmu->base + ARM_SMMU_EVTQ_BASE);
@@ -5573,9 +5590,13 @@ static int __maybe_unused arm_smmu_runtime_suspend(struct device *dev)
 
 	cmd.opcode = CMDQ_OP_CFGI_ALL;
 	arm_smmu_cmdq_issue_cmd_with_sync(smmu, &cmd);
+	if (smmu->options & ARM_SMMU_OPT_TLBI_TWICE)
+		arm_smmu_cmdq_issue_cmd_with_sync(smmu, &cmd);
 
 	cmd.opcode = CMDQ_OP_TLBI_NSNH_ALL;
 	arm_smmu_cmdq_issue_cmd_with_sync(smmu, &cmd);
+	if (smmu->options & ARM_SMMU_OPT_TLBI_TWICE)
+		arm_smmu_cmdq_issue_cmd_with_sync(smmu, &cmd);
 
 	dev_dbg(dev, "Disabling\n");
 	arm_smmu_device_disable(smmu);
diff --git a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.h b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.h
index 9739db5f8a7f..e03a05c671cb 100644
--- a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.h
+++ b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.h
@@ -929,6 +929,7 @@ struct arm_smmu_device {
 #define ARM_SMMU_OPT_MSIPOLL		(1 << 2)
 #define ARM_SMMU_OPT_CMDQ_FORCE_SYNC	(1 << 3)
 #define ARM_SMMU_OPT_TEGRA241_CMDQV	(1 << 4)
+#define ARM_SMMU_OPT_TLBI_TWICE		(1 << 5)
 	u32				options;
 
 	struct arm_smmu_cmdq		cmdq;
-- 
2.43.0

